File: Go/count_negative_numbers_in_a_sorted_matrix.go
/*
1351. Count Negative Numbers in a Sorted Matrix
https://leetcode.com/problems/count-negative-numbers-in-a-sorted-matrix/

Given a m * n matrix grid which is sorted in non-increasing order both row-wise and column-wise.
Return the number of negative numbers in grid.

Example 1:
Input: grid = [[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]]
Output: 8
Explanation: There are 8 negatives number in the matrix.

Example 2:
Input: grid = [[3,2],[1,0]]
Output: 0

Example 3:
Input: grid = [[1,-1],[-1,-1]]
Output: 3

Example 4:
Input: grid = [[-1]]
Output: 1

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 100
-100 <= grid[i][j] <= 100
*/
package main

import (
    "fmt"
)
/*
This solution uses the fact that the negative regions of the matrix will form a "staircase" shape, e.g.:

++++++
++++--
++++--
+++---
+-----
+-----
What this solution then does is to "trace" the outline of the staircase.

Start from bottom-left corner of the matrix, count in the negative numers in each row.
*/
func countNegatives(grid [][]int) int {
    var m int = len(grid)
    var n int = len(grid[0])
    var result int = 0

    var row int = m - 1
    var col int = 0
    for row >= 0 && col < n {
        if grid[row][col] < 0 {
            row -= 1;
            result += n - col
        } else {
            col += 1
        }
    }
    return result
}

File: C++/cpp_code/count_negative_numbers_in_a_sorted_matrix.cpp
/*
https://leetcode.com/problems/count-negative-numbers-in-a-sorted-matrix/description/
1351. Count Negative Numbers in a Sorted Matrix

Given a m x n matrix grid which is sorted in non-increasing order both row-wise and column-wise, return the number of
negative numbers in grid.



Example 1:

Input: grid = [[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]]
Output: 8
Explanation: There are 8 negatives number in the matrix.
Example 2:

Input: grid = [[3,2],[1,0]]
Output: 0


Constraints:

m == grid.length
n == grid[i].length
1 <= m, n <= 100
-100 <= grid[i][j] <= 100


Follow up: Could you find an O(n + m) solution?
*/
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <map>
#include <unordered_set>
#include <algorithm>
#include <cmath>
#include <queue>
#include <stack>
#include <climits>
#include <set>
#include <numeric>
using namespace std;

class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int row = m - 1;
        int col = 0;
        int rst = 0;
        while (row >= 0 && col < n) {
            if (grid[row][col] >= 0) {
                col += 1;
            } else {
                rst += n - col;
                row -= 1;
            }
        }
        return rst;
    }
};